File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week1/Day1/Acwing99/Main.cpp
#include <bits/stdc++.h>

typedef long long LL;
const int N = 5010;

int n, m, k;
int s[N][N];

void solve()
{
   int r;
   std::cin >> n >> r;

   while(n --)
   {
      int x, y, w;
      std::cin >> x >> y >> w;

      x ++, y ++;
      s[x][y] += w;
   }

   r = std::min(r, 5001);
   n = m = 5001;

   for(int i = 1;i <= n;i ++)
      for(int j = 1;j <= m;j ++) {
         s[i][j] += s[i - 1][j] + s[i][j - 1] - s[i - 1][j - 1];
      }

   int res = 0;
   for(int i = r;i <= n;i ++)
      for(int j = r;j <= m;j ++) {
         res = std::max(res, s[i][j] - s[i - r][j] - s[i][j - r] + s[i - r][j - r]);
      }

   std::cout << res << '\n';
}

int main()
{
   std::ios::sync_with_stdio(false);
   std::cin.tie(nullptr);
   std::cout.tie(nullptr);

   int t = 1;
   //std::cin >> t;

   while(t --) solve();
   return 0;
}

File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week1/Day1/Acwing99/main.py
n,r=map(int,input().split())
N=0
a=[[0]*5010 for i in range(5010)]
for i in range(n):
    x,y,w=map(int,input().split())
    N=max(N,x+1,y+1)
    a[x+1][y+1]+=w
res=0
for i in range(1,N+1):
    for j in range(1,N+1):
        a[i][j]+=a[i-1][j]+a[i][j-1]-a[i-1][j-1]
if r<N:
    for i in range(r,N+1):
        for j in range(r,N+1):
            res=max(res,a[i][j]-a[i-r][j]-a[i][j-r]+a[i-r][j-r])
else:
    res=a[N][N]
print(res)

